File: api/Models/Ability.cs
using Microsoft.EntityFrameworkCore;
using System.ComponentModel.DataAnnotations;

namespace api.Models
{
    [Index(nameof(Key), IsUnique = true)]
    public class Ability
    {
        public int Id { get; set; }
        [Required]
        [MaxLength(150)]
        public string Name { get; set; } = default!;
        [Required]
        [MaxLength(150)]
        public string Key { get; set; } = default!;
        [MaxLength(150)]
        public string? Description { get; set; }
        public ICollection<RoleAbility> RoleAbilities { get; set; } = new List<RoleAbility>();
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public DateTime UpdatedAt { get; set; }
        public DateTime? DeletedAt { get; set; }
    }
}

File: api/app/models/ability.ts
import { DateTime } from 'luxon';
import { BaseModel, column, manyToMany } from '@adonisjs/lucid/orm';
import Role from './role.js';
import type { ManyToMany } from '@adonisjs/lucid/types/relations';

export default class Ability extends BaseModel {
  @column({ isPrimary: true })
  declare id: number;

  @column()
  declare name: string;

  @column()
  declare key: string;

  @column()
  declare description: string;

  @column.dateTime({ autoCreate: true })
  declare createdAt: DateTime;

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  declare updatedAt: DateTime;

  @column.dateTime()
  declare deletedAt: DateTime | null;

  @manyToMany(() => Role, {
    pivotTable: 'role_abilities',
    localKey: 'id',
    pivotForeignKey: 'ability_id',
    relatedKey: 'id',
    pivotRelatedForeignKey: 'role_id',
  })
  declare roles: ManyToMany<typeof Role>;
}
